src/main/java/sorald/sonar/SonarLintEngine.java

```diff
@@ -0,0 +1,232 @@
+import org.sonarsource.sonarlint.core.client.api.common.RuleKey;
+import org.sonarsource.sonarlint.core.client.api.common.analysis.DefaultClientIssue;
+import org.sonarsource.sonarlint.core.client.api.common.analysis.IssueListener;
+import org.sonarsource.sonarlint.core.client.api.exceptions.SonarLintWrappedException;
+import org.sonarsource.sonarlint.core.client.api.standalone.StandaloneAnalysisConfiguration;
+import org.sonarsource.sonarlint.core.client.api.standalone.StandaloneGlobalConfiguration;
+import org.sonarsource.sonarlint.core.commons.Language;
+import org.sonarsource.sonarlint.core.commons.log.ClientLogOutput;
+import org.sonarsource.sonarlint.core.commons.progress.ClientProgressMonitor;
+import org.sonarsource.sonarlint.core.commons.progress.ProgressMonitor;
+import org.sonarsource.sonarlint.core.plugin.commons.PluginInstancesRepository;
+import org.sonarsource.sonarlint.core.plugin.commons.PluginInstancesRepository.Configuration;
+import org.sonarsource.sonarlint.core.rule.extractor.SonarLintRuleDefinition;
+import sorald.FileUtils;
+
+public final class SonarLintEngine extends AbstractSonarLintEngine {
+
+    // The order of these initialisations is important as each field is dependent upon the previous
+    // one.
+    private static final String SONAR_JAVA_PLUGIN_JAR_NAME = "sonar-java-plugin.jar";
+    private static final String SONAR_JAVA_PLUGIN_URL =
+            "https://repo1.maven.org/maven2/org/sonarsource/java/sonar-java-plugin/6.12.0.24852/sonar-java-plugin-6.12.0.24852.jar";
+    private static final Path sonarJavaPlugin = getOrDownloadSonarJavaPlugin();
+    private static final StandaloneGlobalConfiguration globalConfig = buildGlobalConfig();
+    private static final PluginInstancesRepositoryWhichCannotBeClosed pluginInstancesRepository =
+            createPluginInstancesRepository();
+    private static final Map<String, SonarLintRuleDefinition> allRulesDefinitionsByKey =
+            computeAllRulesDefinitionsByKey();
+    private static final AnalysisEngineConfiguration analysisGlobalConfig =
+            buildAnalysisEngineConfiguration();
+
+    // The only instance of this singleton class
+    private static SonarLintEngine theOnlyInstance;
+
+    // We need to reinitialise it before starting analysis of any source files on any rules.
+    private AnalysisEngine analysisEngine;
+
+    private SonarLintEngine() {
+        super(null);
+        setLogging(null);
+
+        this.analysisEngine =
+                new AnalysisEngine(analysisGlobalConfig, pluginInstancesRepository, null);
+    }
+
+    private static Path getOrDownloadSonarJavaPlugin() {
+        File cacheDirectory = FileUtils.getCacheDir();
+        String sonarJavaPluginFileName =
+                cacheDirectory + System.getProperty("file.separator") + SONAR_JAVA_PLUGIN_JAR_NAME;
+        File sonarJavaPlugin = new File(sonarJavaPluginFileName);
+        if (sonarJavaPlugin.exists()) {
+            return sonarJavaPlugin.toPath();
```

Comment:
We may need to use reflection to call this method twice and make sure it does not download the Jar the second time.